### ISSUE TITLE
Offer additional note for why method with Self reference cannot be used on `any SomeType`

### ISSUE THREAD

**Issue by ktoso** [swiftlang/swift#76320]
### Motivation

The following snippet;

```

protocol Base {
    associatedtype SomeID: Hashable
}

protocol Alpha: Base {
    func take(id: Self.SomeID)
}

struct Use {
    let alpha: any Alpha
    
    func test() {
        alpha.take(id: 12) // error: Member 'take' cannot be used on value of type 'any Alpha'; consider using a generic constraint instead
    }
}

```

is correct in rejecting the line, since we need `Self.SomeID` to be well typed, but we dont know what its type is via the `any`.

It'd be helpful if we can point out that "why" the method can't be used via a note, in a method with many parameters one can sometimes miss the Self reference.

### Proposed solution

Adding a `note` on the offending parameter would be good

**Comment by amritpan** [comment]
@GunjanRawat26 This could be a good intro to diagnostics

**Comment by ktoso** [comment]
Linking with radar rdar://135445764

Contributions welcome here, this could be an simple first issue maybe?

**Comment by flash1729** [comment]
hey @ktoso , i would like to work on this as my first issue

from what i have understood is we need to give a more refined compiler message for this error and for it
1. identify where in the compiler this error message is generated
2. change the message there to be more specific

it would be great if you could guide me further on anything that i am missing

**Comment by flash1729** [comment]
Current progress : 
I found the code related to error message in DiagnosticsSema.def 
```swift
ERROR(could_not_use_member_on_existential,none,
      "member %1 cannot be used on value of type %0; consider using a"
      " generic constraint instead",
      (Type, DeclNameRef))

// error: member 'take' cannot be used on value of type 'any Alpha'; consider using a generic constraint instead
```
This is where changes will be made to show different error message

### Questions

1. Are there any related places where this code is rewritten or referenced so changes will be made there as well ?
2. this is my first contribution so i dont know how to set up environment to test our changes so for that i am reading [GettingStarted.md](https://github.com/swiftlang/swift/blob/main/docs/HowToGuides/GettingStarted.md). I am a bit unsure about exactly what needs to be done rn.
3. I am understanding what error is exactly. also are there any specific guidelines for wording error message ?

**Comment by Lenhart94** [comment]
protocol Base {
    associatedtype SomeID: Hashable
}

protocol Alpha: Base {
    func take(id: Self.SomeID)
}

struct Use<A: Alpha> where A.SomeID == Int {
    let alpha: A
    
    func test() {
        alpha.take(id: 12)
    }
}

**Comment by Lenhart94** [comment]
protocol Base {
    associatedtype SomeID: Hashable
}

protocol Alpha: Base {
    func take(id: Self.SomeID)
}

struct Use<A: Alpha> where A.SomeID == Int {
    let alpha: A
    
    func test() {
        alpha.take(id: 12)
    }
}

**Comment by jameesbrown** [comment]
@flash1729 I know you asked the original developer for help, and I’m sure they’ll provide guidance when they’re available. In the meantime, I’m very new myself, but I’ll try my best to give some direction. 

Based on the Proposed Solution in the issue submission, it seems like adding a note instead of modifying the existing error would be a good approach. Here's what I would try (though you might explore other ideas too):

1. To locate all references to could_not_use_member_on_existential, you can use `grep -r "could_not_use_member_on_existential" path/to/swift-project/swift` or search manually using CMD+F in Xcode.

2. For initial debugging, you can pass the `-typecheck -verify` arguments, and then locate the relevant test cases. A good way to find related tests for this error would be to run `grep -rl "cannot be used on value of type" path/to/swift-project/swift/test`. Additionally, you may find the documentation [here](https://github.com/swiftlang/swift/blob/main/docs/Diagnostics.md) helpful.

3. You might also want to review this document: [Existential Member Access Limitations](https://github.com/swiftlang/swift/blob/main/userdocs/diagnostics/existential-member-access-limitations.md). Looking at how other notes are worded across the codebase could give you insights into how to phrase yours.

I hope this helps! Let me know if I’m overstepping or if anything needs more explanation.

**Comment by ktoso** [comment]
That's good hints, thanks @jameesbrown

**Comment by flash1729** [comment]
@jameesbrown Thanks for hints. I will look into them...
For now what i did was i built swift in my device for first time and then after making changes to DiagnosticsSema.def if rebuilt and tried to compile a test file using the modified compiler and as expected i got the Modified Error message which is sort of like this
```
member %1 cannot be used on value of type %0; the method requires knowledge of the concrete type for 'Self.SomeID'. Consider using a generic constraint instead
```
I faced issues in determining which schema to use (i used **diagnostics_database**)
and later couldn't completely get the hang of which builds to do to ensure complete build is good after changes so if you can suggest some additional reading related to this then it would be great...
Also unknowingly i used two build methods (ninja (for complete build) and xcode(after minor changes)) which led to some errors cause xcode build couldn't find some files in expected location of llvm clone. 

Now i will go through suggested sources and instead of modifying error message add a related note accordingly...

**Comment by jameesbrown** [comment]
@flash1729 
> I faced issues in determining which schema to use (i used **diagnostics_database**)
> and later couldn't completely get the hang of which builds to do to ensure complete build is good after changes so **_if you can suggest some additional reading related to this then it would be great..._**

You may have already looked at it, but https://github.com/swiftlang/swift/blob/main/docs/HowToGuides/GettingStarted.md#using-ninja-with-xcode is the best. 

Also one thing to be careful about is when selecting the executable for your scheme (Edit Scheme...>Run), make sure you are selecting one in `build/Ninja-*/swift-macosx-*/bin`, the ones that appear in the dropdown (for me at least) are all `build/Xcode-*/swift-macosx-*/bin` and will not work. If this is the same case for you, then you can select: Other... and choose the executable via finder.

**Comment by Anshika9gaur** [comment]
Hi! I'd like to explore this issue and see if I can contribute a solution. I'll share my progress soon. If anyone has suggestions or insights, they’re welcome!

**Comment by Niyatijain-9** [comment]
Hi, would like to work on this. Could you  assign this to me.

**Comment by flash1729** [comment]
Yes. Due to some other commitments I was unable to submit a final pr.
I'll look into it and submit it 👍🏻

**Comment by SwiftyJoeyy** [comment]
Hey @flash1729 are you still working on this issue? If not, I'd like to take it over.

**Comment by AnthonyLatsis** [comment]
To set a reasonable goal for this good first issue, there is no need to take the solution as far as to state which part of the protocol member’s type signature is the cause. Just mentioning that the member references either `Self` or a particular associated type would already be a very helpful improvement.

**Comment by flash1729** [comment]
@SwiftyJoeyy please do look into this issue.
I wont be able to open a pr anytime very soon so its best if you can do so 👍

**Comment by SwiftyJoeyy** [comment]
Hey @ktoso I'm currently working on this, and I'd like to check if I'm on the right path, or if there's a simpler way to approach the problem—especially after encountering some very complex test cases in [test/decl/protocol/existential_member_access/concrete.swift](https://github.com/swiftlang/swift/blob/ff790c6d01edab1695c744d22dad0d8b32f2d749/test/decl/protocol/existential_member_access/concrete.swift).

In [CSDiagnostics.cpp](https://github.com/swiftlang/swift/blob/main/lib/Sema/CSDiagnostics.cpp#L4703), I’m performing member lookup using `ConstraintSystem::performMemberLookup` and, if the candidate is invalid, I try to walk through the type to find all references to `Self` or `DependentMemberType`. For each of these, I emit a note to clarify why the member can't be used.

To extract references to `Self` or associated types, I use this recursive function to walk the type returned by `getInterfaceType()`:

```
void insertUnresolvedTypesFromRec(Type type, SmallSetVector<Type, 2> &matches) {
  if (auto genFuncTy = dyn_cast<GenericFunctionType>(type)) {
    auto result = genFuncTy->getResult();
    insertUnresolvedTypesFromRec(result, matches);
    return;
  }
  
  if (auto funcTy = dyn_cast<FunctionType>(type)) {
    auto parameters = funcTy->getParams();
    for (auto param : parameters) { // Func params
      auto ty = param.getParameterType();
      insertUnresolvedTypesFromRec(ty, matches);
    }
    
    insertUnresolvedTypesFromRec(funcTy->getResult(), matches); // Func return type
    return;
  }
  
  if (auto *paramTy = dyn_cast<GenericTypeParamType>(type.getPointer())) {
    if (paramTy->getDepth() == 0 && paramTy->getIndex() == 0) { // Self
      matches.insert(type);
    }
  }else if (type->getKind() == TypeKind::DependentMember) { // Associated type
    matches.insert(type);
    
  }else if (auto boundTy = dyn_cast<BoundGenericType>(type)) {
    for (auto ty : boundTy->getGenericArgs()) {
      insertUnresolvedTypesFromRec(ty, matches);
    }
  }
}
```

This works for many cases, but I’ve noticed that tests in existential_member_access/concrete.swift have some extremely complex combinations of `Self`, associated types, and generics. Handling all of these correctly adds a lot of complexity to the type walker logic. 

Before I go too deep into trying to handle all of that, I wanted to ask:

1. Is this kind of recursive type walking a reasonable direction?
2. Is there a simpler or more canonical way in the Swift codebase to extract dependent member type references from a declaration's interface type?
3. Should I perhaps limit the diagnostic improvement to clearer/common cases and avoid over engineering support for extremely rare constructs?

Any advice would be very appreciated. This is my first time trying to contribute to Swift, so thanks in advance for your time and guidance!

**Comment by ktoso** [comment]
Thanks for picking this up! Maybe @slavapestov would be better suited to answer these questions though?

**Comment by SwiftyJoeyy** [comment]
Hey @slavapestov could you please help us with this? Thank you 🙏
